### PR TITLE
Upgrade go and 3rd party libraries and improve the tests

### PR DESCRIPTION
**Pull request by renato-araujo-form3**
## Context

The current go Go version of the tool used is version 1.14, which is old since we are on version 1.18.
Some of the libraries used by the tool are in their old versions, providing limitations in relation to their current APIs (for example fetching the user timezone).

The current solution does not have a large test coverage, covering only the configuration, which increases the complexity and reduces the confidence of introducing new changes in the tool (for example, removing the limitation of generating reports based on different timezones).

This PR introduces Go and external dependency upgrades together with an increase of testing coverage.

## Proof of Work

Some changes were introduced in this PR, such as:

### API Package
- Removing dependencies of low-level implementations of PagerDuty library functions, and introducing an interface for high-level abstractions. This facilitated the tests, which enable to introduce the usage of mocks, of the methods in the API package and increased the test coverage from 0 to 96.7%.
  - Screenshot 2022-08-25 at 12 55 31
- Separated the files per context to facilitate de tests and improve the readability of the code. 
- Created explicit data structures which are converted from the data provided by the PagerDuty library to give us more control over the tests and necessary data to generate the expected results from the tool.

### Command (CMD) package
- Removing dependencies of low-level implementations of API package functions, and introducing an interface for high-level abstractions.
  - It does not remove the dependency of the API package as it depends of the data structures.
  - This was implemented to facilitate to use mocks and create some tests for the commands, the tests only validate the dependency of the API package, and it requires more work to cover the commands fully. To be able to tests the commands some changes in the output are necessary, which can change the logic of the application, in this case it's not in the context of this PR.
  - The coverage was increased by 14.3%, and it didn't cover the generate_repot.go file, which has a large complexity and it will be covered when introducing the new changes (for example extending the timezone support). 
  - Screenshot 2022-08-25 at 13 17 45
 
### The generated report
- The report generation still needs to be checked manually as it's complexity still requires some refactoring and validation of the business rules. 
- The test variables:
  - Considering from `01 Sep 22 00:00 UTC` to `01 Oct 22 00:00 UTC`;
  - Prices configured for `£1` to weekday and `£2` for weekend and bank holidays;
  - 1 schedule created to exclude from remuneration considering only 09:00 BST to 17:00 BST;
  - Screenshot 2022-08-25 at 13 34 01
  - 1 Schedule created to include in the remuneration considering weekdays from 17:00 BST to 09:00 BST of the other day, and 09:00 BST to 09:00 BST for every weekend day;
  - Screenshot 2022-08-25 at 13 35 17

The final report:
Screenshot 2022-08-25 at 13 38 26

For more reference, below you can check the report generated by the current version in master using the same parameters.
[pagerduty_oncall_report.8-2022.pdf](https://github.com/form3tech-oss/go-pagerduty-oncall-report/files/9432034/pagerduty_oncall_report.8-2022.pdf)

### CODE REVIEW

File: api/pagerduty.go [hunk 1]
```diff
@@ -8,17 +8,17 @@ import (
 
 var Client *PagerDutyClient
 
-type PagerDutyClient struct {
-	apiClient *pagerduty.Client
+type PdClient interface {
+	ListSchedules(o pagerduty.ListSchedulesOptions) (*pagerduty.ListSchedulesResponse, error)
```
Comment (zhenya-khvan-form3) [inline api/pagerduty.go:12]: I think adding these interfaces and using mocks is a great stepping stone to having better code coverage! 🎉 

I have the opinion that the **next** step would be to leverage a "fake" server that imitates the API endpoints of pagerduty. I think that would provide better resiliency in refactoring because we're only binding ourselves to the actual HTTP interface, not the `go` interface.

I say this because I notice that all these endpoints without `Context` are actually deprecated, so in order to change the codebase to use the `WithContext` variants of the methods, we will have to refactor the tests (which always seem like a smell 💨).

File: cmd/list_schedules_test.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+func Test_listSchedules(t *testing.T) {
+	tests := []struct {
+		name      string
+		mockSetup func(*clientMock)
+		wantErr   bool
+	}{
+		{
+			name: "Successfully list pagerduty schedules",
```
Comment (zhenya-khvan-form3) [inline cmd/list_schedules_test.go:20]: I'm kinda confused about the difference between the tests in the `cmd` project vs the `api` project.

For example, it seems like here is essentially a duplicate (with less assertions) of the test here:

https://github.com/form3tech-oss/go-pagerduty-oncall-report/pull/38/files#diff-62820d979540a4fb204e4867f977d19f47720a3d5a757332b1e0809ada9bcdc3R29

I think if we could combine them somehow, it would provide a more wholistic test and reduce some copy-paste between `cmd` and `api` tests?

**Review by zhenya-khvan-form3 — DISMISSED** [review]
Looks great, I think improving the test coverage of the tool provides much greater confidence in being able to change the codebase!

**Review by jordan-van-dyk-form3 — APPROVED** [review]

File: api/pagerduty.go (hunk 1 above)
Comment (renato-araujo-form3) [reply api/pagerduty.go:12]: That is something that we could cover for the future when connecting them with problems that we really want to solve. 

I realised that the functions were deprecated, and the only reason I didn't introduce the changes by fixing them was because I would need to change the application logic, which I avoided as much as possible.

File: api/mock_test.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
+// Code generated by mockery v2.14.0. DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (renato-araujo-form3) [reply api/mock_test.go:1]: The content of the file is actually generated but the name of the file was change by me, I will fix that.

File: cmd/root.go [hunk 4]
```diff
@@ -27,27 +38,22 @@ func init() {
...
-	} else {
-		// Find home directory.
+	if cfgFile == "" {
 		home, err := homedir.Dir()
 		if err != nil {
-			log.Fatal("Can't get the homedir: ", err)
+			log.Fatal("failed to get home directory: ", err)
 		}
 
-		viper.AddConfigPath(home)
```
[the remark this answers is not in the document]
Comment (renato-araujo-form3) [reply cmd/root.go:42]: The intention of the PR is actually to not introduce any change that potentially affects the behaviour of the application and if so, cover with tests.
Having this in mind I will undo this change.

File: cmd/list_schedules_test.go (hunk 2 above)
Comment (renato-araujo-form3) [reply cmd/list_schedules_test.go:20]: These tests were introduced by just to cover a small expectation, not perfect as described in the PR.
The intention of it was just to check if the dependency would be satisfied removing the global `api.Client`.

File: api/pagerduty.go (hunk 1 above)
Comment (zhenya-khvan-form3) [reply api/pagerduty.go:12]: Yeah agree, was just something to mention for later down the line 🕐

**Review by zhenya-khvan-form3 — APPROVED** [review]

File: api/services.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+func (p *PagerDutyClient) ListServices(teamID string) ([]*Service, error) {
+	var opts pagerduty.ListServiceOptions
+	opts.TeamIDs = []string{teamID}
+	listServicesResponse, err := p.ApiClient.ListServices(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	var serviceList []*Service
+	for _, service := range listServicesResponse.Services {
```
Comment (wojciecholszewski-form3) [inline api/services.go:19]: Don't we need to handle pagination here similarly we do in ListSchedules https://github.com/form3tech-oss/go-pagerduty-oncall-report/pull/38/files#diff-993fa9b341588b541152084e056f19fcf69b69da4c7a0943368bb3b73991d76cR40?

File: api/teams.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+func (p *PagerDutyClient) ListTeams() ([]*Team, error) {
+	var opts pagerduty.ListTeamOptions
+	listTeamsResponse, err := p.ApiClient.ListTeams(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	var teamList []*Team
+	for _, team := range listTeamsResponse.Teams {
```
Comment (wojciecholszewski-form3) [inline api/teams.go:18]: no need to handle pagination here?

File: cmd/list_schedules.go [hunk 7]
```diff
@@ -4,22 +4,30 @@ import (
...
 	Short: "list schedules on PagerDuty",
 	Long:  "Get the list of schedules configured in PagerDuty",
-	RunE:  listSchedules,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		pd := &pagerDutyClient{client: api.NewPagerDutyAPIClient(Config.PdAuthToken)}
+		err := pd.listSchedules()
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (wojciecholszewski-form3) [inline cmd/list_schedules.go:21]:
```suggestion
		return pd.listSchedules()
```

isn't it easier to return err right away?

File: cmd/list_services.go [hunk 8]
```diff
@@ -12,16 +13,22 @@ var listServicesCmd = &cobra.Command{
...
 	Long:  "Get the list of services configured in PagerDuty",
 	Args:  cobra.ExactArgs(1),
-	RunE:  listServices,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		pd := &pagerDutyClient{client: api.NewPagerDutyAPIClient(Config.PdAuthToken)}
+		err := pd.listServices(args[0])
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (wojciecholszewski-form3) [inline cmd/list_services.go:22]:
```suggestion
		return pd.listServices(args[0])
```
similar remark to the one I posted above
if we don't have any special logic for handling error why not return right away?

File: cmd/list_services_test.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
+package cmd
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/form3tech-oss/go-pagerduty-oncall-report/api"
+
```
Comment (wojciecholszewski-form3) [inline cmd/list_services_test.go:8]: nit: redundant empty line

File: go.mod [hunk 10]
```diff
@@ -1,32 +1,40 @@
...
+	github.com/PagerDuty/go-pagerduty v1.5.1
+	github.com/jung-kurt/gofpdf v1.0.0
+	github.com/mitchellh/go-homedir v1.1.0
+	github.com/spf13/cobra v1.5.0
+	github.com/spf13/viper v1.12.0
+	github.com/stretchr/testify v1.8.0
+	gopkg.in/yaml.v2 v2.4.0
+)
+
+require (
```
Comment (wojciecholszewski-form3) [inline go.mod:16]: what is the reason behind having 2 `require` blocks? cannot they be merged into single one?

**Review by wojciecholszewski-form3 — COMMENTED** [review]
Few comments left. Nothing major though.

File: cmd/list_services_test.go (hunk 9 above)
Comment (renato-araujo-form3) [reply cmd/list_services_test.go:8]: This is actually a good practice in the go world, when you separate your dependencies in different blocks, having in the structure the separation between native, internal and external packages.

You can see that reference in some articles in the internet, for example:
- [Importing Packages in Go by DigitalOcean](https://www.digitalocean.com/community/tutorials/importing-packages-in-go)
- [How to Use Go Modules by DigitalOcean](https://www.digitalocean.com/community/tutorials/how-to-use-go-modules)

File: cmd/list_services.go (hunk 8 above)
Comment (renato-araujo-form3) [reply cmd/list_services.go]: fixed

File: cmd/list_schedules.go (hunk 7 above)
Comment (renato-araujo-form3) [reply cmd/list_schedules.go]: fixed

File: cmd/list_teams.go [hunk 11]
```diff
@@ -4,22 +4,30 @@ import (
...
 	Short: "list teams on PagerDuty",
 	Long:  "Get the list of teams configured in PagerDuty",
-	RunE:  listTeams,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		pd := &pagerDutyClient{client: api.NewPagerDutyAPIClient(Config.PdAuthToken)}
+		err := pd.listTeams()
+		if err != nil {
+			return err
+		}
+		return nil
```
[the remark this answers is not in the document]
Comment (renato-araujo-form3) [reply cmd/list_teams.go]: fixed

File: api/teams.go (hunk 6 above)
Comment (renato-araujo-form3) [reply api/teams.go:18]: This is definitely something that we want to address in the future because it is a potential issue, but we need to be certain that the problem is limiting us and it will require further investigation. The reason that is not addressed now, is because that would introduce new point of failure changing the logic of the application which is not the context of the PR and should be covered with tests.

File: go.mod (hunk 10 above)
Comment (renato-araujo-form3) [reply go.mod:16]: This was introduced to Go in version 1.17 to separate the direct and indirect dependencies. You can see the second block has only indirect dependencies. 

If you want to go deep on understanding the dependency graph has changes you can check the official documentation: https://go.dev/doc/go1.17#go-command

TL;DR;
> Because the number of explicit requirements may be substantially larger in an expanded Go 1.17 go.mod file, the newly-added requirements on indirect dependencies in a go 1.17 module are maintained in a separate require block from the block containing direct dependencies.

File: api/services.go (hunk 5 above)
Comment (renato-araujo-form3) [reply api/services.go:19]: See: https://github.com/form3tech-oss/go-pagerduty-oncall-report/pull/38#discussion_r959449759

File: cmd/list_services_test.go (hunk 9 above)
Comment (wojciecholszewski-form3) [reply cmd/list_services_test.go:8]: Didn't know that. Thank you for sharing 👍

**Review by wojciecholszewski-form3 — APPROVED** [review]

**Review by zhenya-khvan-form3 — APPROVED** [review]

**Review by jordan-van-dyk-form3 — APPROVED** [review]
